Restore theme after closing/reopening the app.

vishwesh3 opened the pull request:
Fix:#505

**Summary of changes**

Save preferred theme in Shared Preferences. And restore them on reopening app.

Screenshots or a brief video showing the change in action:

![ezgif com-video-to-gif-2](https://cloud.githubusercontent.com/assets/18511177/26751891/903097a6-4861-11e7-9389-8bca910bdd0d.gif)

Please make sure these boxes are checked before submitting your pull request - thanks! [Guide](./PULL_REQUEST_GUIDE.md)

- [x] Code is formatted.

- [x] Run the lint tests with `./gradlew lintDebug` and make sure BUILD is SUCCESSFUL

- [x] If new feature is implemeted then it is compatible with Night theme too.

- [x] Commit messages are well-written.

kunall17 commented:
@timabbott @vishwesh3 @Sam1301 should we use the auto night theme, it changes the theme according to the time going on?

https://developer.android.com/reference/android/support/v7/app/AppCompatDelegate.html#MODE_NIGHT_AUTO

Sam1301 commented:
Nice idea!, though I am somewhat biased to user having control over the theme regardless of time, but this is debatable.

saketkumar95 commented:
I guess No! We should keep the preference of the User only. Twitter app flows the same flow, i.e It all depends on the user choice. But It's nice idea. Some apps like Google maps uses this. :) 
@kunall17 Lets checkout what Slack's follows..  :D

kunall17 commented:
Yeah but imagine a situtation where I enabled the night theme around 9 PM, and then I started the app in the next morning I would prefer, that it would change automatically to the day mode in the morning! 

One solution is we can have a sub menu like the Filter menu, and have a checkbox there for auto changing and one option for changing the current theme?

kunall17 commented on app/src/main/java/com/zulip/android/ZulipApp.java:
```diff
@@ -598,4 +598,15 @@ public void setMessageContentEditParams(int seconds, boolean param) {
+     * Update preferred theme
+     * @param b true if night theme is preferred
+     */
+    public void makeNightThemeDefault(boolean b) {
```
Make this boolean b something more understandable!

vishwesh3 replied:
Done 👍

kunall17 commented on app/src/main/java/com/zulip/android/util/Constants.java:
```diff
@@ -27,4 +27,5 @@
     //if two continuous messages are from same sender and time difference is less than this then hide it
     public static long HIDE_TIMESTAMP_THRESHOLD = 60 * 1000;// 1 minute
     public static String ACTION_MESSAGE_PUSH_NOTIFICATION = "ACTION_MESSAGE_PUSH_NOTIFICATION";
+    public static String NIGHT_THEME = "isNightTheme";
```
Let's follow the convention going on, `public static String NIGHT_THEME = 'NIGHT_THEME'`

vishwesh3 replied:
Done 👍

kunall17 commented on app/src/main/java/com/zulip/android/activities/ZulipActivity.java:
```diff
@@ -2358,12 +2365,15 @@ public void onClick(
                         setNightMode(AppCompatDelegate.MODE_NIGHT_YES);
+                        app.makeNightThemeDefault(true);
                         break;
                     case AppCompatDelegate.MODE_NIGHT_YES:
```
NIT: As the default and this is same, we can remove the block (Upto you how you want to do it) 
```
setNightMode(AppCompatDelegate.MODE_NIGHT_NO);
app.makeNightThemeDefault(false);
```

vishwesh3 replied:
Done 👍 . Removed block.

vishwesh3 commented:
Updated 👍 

Added functionality for auto night theme.
- Added a checkbox option in menu.
- If auto theme is enabled then irrespective of preffered theme, it
will be overridden by auto theme (depending on time).
- User can switch theme even if auto theme is enabled, but it is for that instance
only.

kunall17 commented:
Thanks for the quick awesome work @vishwesh3.
Waiting for confirmation of @timabbott to merge this PR, after that we can do a beta release again.

timabbott commented:
I think this seems reasonable.  One the question of whether to do auto-night, I think long-term maybe we want 3 setting values for this sort of feature: day/night/auto, so the user can choose how they want it to work.

vishwesh3 commented:
@timabbott Added 3 options.
![3464974478807180902 account_id 1](https://cloud.githubusercontent.com/assets/18511177/26764362/526e1b5a-4983-11e7-9b1a-c2cc29aaeffe.jpg)
![3757914563653689949 account_id 1](https://cloud.githubusercontent.com/assets/18511177/26764361/525b53da-4983-11e7-916a-3c5b9cdfb66a.jpg)

Sam1301 commented:
nice! we can probably omit "theme" in each option.

vishwesh3 commented:
@Sam1301 Updated 👍 
![screenshot_1496639437](https://cloud.githubusercontent.com/assets/18511177/26771421/bb5d9390-49db-11e7-8d21-954c5bdefcd5.png)

kunall17 commented:
Cool this looks awesome! 👍

kunall17 commented:
Merged #506 👍
